### DISCUSSION TITLE
How can we use our trained model as pretrained model in MONAILabel (3D Slicer)

### DISCUSSION THREAD

**Discussion by Suchi97** [Project-MONAI/MONAILabel#697] [General]
Hi Team, 

We are trying to use our trained model to predict the labels in MONAI Label via 3D Slicer. Our trained model is trained on the labels background and spleen and we are trying to use this trained model as pretrained model and then predict the spleen using MONAI Label via 3D silcer. So, this model having a mean dice of 0.95. Since this model having input channels as 1 and output channels as 2,  we changed the input from 1 to 3 (background, spleen, image) and output channels to  2 (background and spleen) to match the dimensions. When I use this trained model, the prediction result looks different. Spleen organ is not correctly predicted.

So, how can we use our trained model as the pretrained model in MONAI Label which gives the same result when predicted outside the MONAI Label.

Can you help me to understand how the pretrained model used in MONAI Label is trained. Like, is the model is trained using MONAI Label via 3D Slicer or any other methods. 

Any suggestions is appreciated.
Thanks in advance

**Comment by diazandr3s** [comment]
Hi @Suchi97,

This is a good question. Which App are you using? How did you get the pre-trained model? 

Importing pre-trained models obtained from other pipelines is something MONAI Label can be used for.

My suggestion is to use the Segmentation App. Other Apps such as DeepGrow and DeepEdit have more sophisticated training loops that involved click simulation. Input tensors of these Apps have several channels as you may know: image, foreground, and background clicks.

As the Segmentation App is a non-interactive model, the input tensor is only the image.

The pre-trained model should be located in a folder called model inside the folder App and the .pt file should be named [pretrained](https://github.com/Project-MONAI/MONAILabel/blob/main/sample-apps/segmentation/main.py#L59)

You should also make sure that the network architecture is the same as the pre-trained model https://github.com/Project-MONAI/MONAILabel/blob/main/sample-apps/segmentation/main.py#L48


Please let us know how that goes.

**Reply by Suchi97** [reply]
Thanks @diazandr3s  for your reply. 

Details of our custom trained model : We trained our model on spleen organ. This model has the input channel as 1 and output channel as 2 (background, spleen).  And the model has a dice of 0.95. 

Our aim is to use our own custom trained model as a pretrained model in DeepEdit app using MONAI Label 

To do this, what all changes should be done to the trained model. We can see that the deepedit app is an interactive model. So, how can we use a custom pretrained model in deepedit /deepedit_multilabel app.

---

As per your comments, we used the segmentation app. 
When we run the segmentation app (without changing any statements), we ran into some issues related to mmar. This issue is solved by using **RemoteMMARKeys**. 

So we updated `MONAILabel/monailabel/interfaces/app.py` and `MONAILabel/sample-apps/segmentation_spleen/main.py`. Then we could able to run the MONAI Label server and obtained the below result.

![Screenshot (89)](https://user-images.githubusercontent.com/25792114/159465255-7b3874e5-b471-4044-b12d-061b80ff329c.png)

Then I used our trained model (spleen) as the pretrained model. This also works fine. Even though the model has a mean dice of 0.95, the model didn't predict correctly. Is there anything more to change in the segmentation_spleen app?

The predicted image looks like as below:

![Screenshot (90)](https://user-images.githubusercontent.com/25792114/159465888-d006d0c8-d092-4a7b-a06c-7bc541840ab8.png)


Thanks again for your help

**Reply by diazandr3s** [reply]
Thanks for your response, @Suchi97.
I'm glad to see the Segmentation App works using your pre-trained model.
With regards to this point:

> Our aim is to use our own custom trained model as a pretrained model in DeepEdit app using MONAI Label
> To do this, what all changes should be done to the trained model. We can see that the deepedit app is an interactive model. So, how can we use a custom pretrained model in deepedit /deepedit_multilabel app.

As the DeepEdit is an interactive app, the pre-trained model you're using should have to be trained with simulated clicks in the training loop. 

From what you're saying, your pre-trained model wasn't trained to be interactive. This means you can't use that pre-trained model with the DeepEdit App.

However, one of the great things about MONAI Label is that it facilitates the training of complex models such as DeepEdit.

My suggestion then will be this: 1/ you use DeepEdit to get a pre-trained model or 2/ work with the Segmentation App to use your pre-trained model.

Regarding the results you're getting using the Segmentation App, please make sure you apply the same pre-transforms in infer (https://github.com/Project-MONAI/MONAILabel/blob/main/sample-apps/segmentation/lib/infer.py#L72).  

I wouldn't be surprised anyway if the pre-trained model needs update/fine-tuning. A mean dice of 95% doesn't mean much if the number and characteristics of the validation set do not represent the test set.

Hope this helps.

**Reply by Suchi97** [reply]
Thanks @diazandr3s  for your detailed information.

**Reply by Suchi97** [reply]
Thanks `@diazandr3s` again for your help. 
I added the pre transforms that we used when training the images in the `infer.py`.
Now, our trained model also predicted the spleen organ correctly.

![image](https://user-images.githubusercontent.com/25792114/159666024-8a108cb8-ec4b-4cd6-8960-d20759a77835.png)

**Reply by SachidanandAlle** [reply]
Look for latest example (more easier for anyone to add their own models and use them)
https://github.com/Project-MONAI/MONAILabel/tree/main/sample-apps/radiology#how-to-add-new-model

You can install monailabel from github or latest docker or https://pypi.org/project/monailabel/0.4.0rc1/

and you start monailabel as `monailabel start_server -a radiology -s images -c models <your_model>`

**Reply by Suchi97** [reply]
Thanks @SachidanandAlle for this new information. I will try this and let u know the results.

**Reply by diazandr3s** [reply]
Superb! Glad to hear this, @Suchi97  :)

**Reply by drchen123** [reply]
> Hi @Suchi97,
> 
> This is a good question. Which App are you using? How did you get the pre-trained model?
> 
> Importing pre-trained models obtained from other pipelines is something MONAI Label can be used for.
> 
> My suggestion is to use the Segmentation App. Other Apps such as DeepGrow and DeepEdit have more sophisticated training loops that involved click simulation. Input tensors of these Apps have several channels as you may know: image, foreground, and background clicks.
> 
> As the Segmentation App is a non-interactive model, the input tensor is only the image.
> 
> The pre-trained model should be located in a folder called model inside the folder App and the .pt file should be named [pretrained](https://github.com/Project-MONAI/MONAILabel/blob/main/sample-apps/segmentation/main.py#L59)
> 
> You should also make sure that the network architecture is the same as the pre-trained model https://github.com/Project-MONAI/MONAILabel/blob/main/sample-apps/segmentation/main.py#L48
> 
> Please let us know how that goes.

Dear sir
    I am a vascular surgeon, i have a high volume of CTA data, and i need to train my own model eventually for auto-segmentation of the aorta for different disease (abdominal aortic aneurysm, aortic dissection, intramural hematoma), i also want to public these trained models, so other clinicians major in these diseases could use it. 
    The methods you provide are of little difficult for me, could you provide a step-to-step tutorial to create my own model based on existing files in radiology? 
    Thanks, and look forward to hear from you.

**Reply by diazandr3s** [reply]
Hi @drchen123,

We're glad you consider MONAI Label for your task. Let's chat more about this via email or a videocall.
Please send me an email (adiazpinto@nvidia.com) so we further discuss this.

**Comment by Suchi97** [comment]
Hi Team, 

We trained a model having the labels **background**, **spleen** and **liver**. For the training purpose, we used **BTCV dataset.** 

We are trying to use this model as a pretrained model in MONAI Label server. Here, I used the **Segmentation app**.

To do this, we followed the below steps:

Changes done in `MONAI_Label_Multilabel/MONAILabel/sample-apps/segmentation/main.py`

1. Updated the labels dictionary to  `self.label_names = { "spleen": 1,  "liver": 2,  "background": 0, }`
2. Changed the spatial size to which we used for training
3. Copied our trained model to `MONAILabel/sample-apps/segmentation/model` directory.
4. Changed the pretrained model name to our trained model name.

Changes done in `MONAILabel/sample-apps/segmentation/lib/infer.py`
1. Updated the pre-transforms. 
```           
              LoadImaged(keys=["image"]), 
              EnsureChannelFirstd(keys=["image"]),
              ScaleIntensityRanged(keys=["image"], a_min=-175, a_max=250, b_min=0, b_max=1.0, clip=True),
              EnsureTyped(keys=["image"],device=data.get("device") if data else None)
``` 
These are the pretransforms which I used for training purpose.

When I inferred a sample image of BTCV dataset outside MONAI Label, we got the below result

![image](https://user-images.githubusercontent.com/25792114/159949907-f1cdbad1-9a90-4b74-b49a-7e4fe1921801.png)

From the above image, we can see that the predicted results are almost good. After training we got a mean dice of 0.93.

When we used the same model in 3D Slcier, the below results are obtained.

![image](https://user-images.githubusercontent.com/25792114/159950282-36a2d332-ca39-4531-bbc7-eeafd8d023f4.png)

From the above image, we can see that the results are not matched correctly.

Can you please help me to solve this

Any suggestions is appreciated.
Thanks in advance :)

**Comment by diazandr3s** [comment]
Hi @Suchi97,

Thanks for your comment. It is nice to see the results you're getting.
I think the problem comes from the inverse transforms. I need to update that bit in the segmentation app. 
Please [add these lines](https://github.com/Project-MONAI/MONAILabel/blob/0.3.1/sample-apps/deepedit_multilabel/lib/infer.py#L92-L93) in the infer.py file in the Segmentation App. That should solve the problem.

Please let us know how that goes.

Thanks again!!

**Reply by Suchi97** [reply]
Thanks @diazandr3s for your reply.
Thanks again for your suggestions. Beacuse of that we could able to see the progress :)

[These lines ](https://github.com/Project-MONAI/MONAILabel/blob/0.3.1/sample-apps/deepedit_multilabel/lib/infer.py#L92-L93) are already present in the **Segmentation app** ``infer.py ``.

![image](https://user-images.githubusercontent.com/25792114/159964202-88d32ba1-a42a-46d5-b27b-d708f0c444fa.png)

**Reply by diazandr3s** [reply]
> Thanks @diazandr3s for your reply. Thanks again for your suggestions. Beacuse of that we could able to see the progress :)
> 
> [These lines ](https://github.com/Project-MONAI/MONAILabel/blob/0.3.1/sample-apps/deepedit_multilabel/lib/infer.py#L92-L93) are already present in the **Segmentation app** `infer.py `.
> 
> ![image](https://user-images.githubusercontent.com/25792114/159964202-88d32ba1-a42a-46d5-b27b-d708f0c444fa.png)

I'm a bit confused ... the link you sent me is to the deepedit multilabel ... I meant the pure Segmentation App (https://github.com/Project-MONAI/MONAILabel/blob/0.3.1/sample-apps/segmentation/lib/infer.py#L60) 

Is it the one you're using, right?

**Reply by Suchi97** [reply]
Yes @diazandr3s , we are  using segmentation app. 
Sorry, I was also confused because in the below comment the link refers to the deepedit_multilabel

> Thanks @diazandr3s for your reply. Thanks again for your suggestions. Beacuse of that we could able to see the progress :)
> 
> [These lines ](https://github.com/Project-MONAI/MONAILabel/blob/0.3.1/sample-apps/deepedit_multilabel/lib/infer.py#L92-L93) are already present in the **Segmentation app** `infer.py `.
> 
> ![image](https://user-images.githubusercontent.com/25792114/159964202-88d32ba1-a42a-46d5-b27b-d708f0c444fa.png)



And `MONAILabel/sample-apps/segmentation/lib/infer.py` also has the above mentioned function
![image](https://user-images.githubusercontent.com/25792114/159982269-bed7195b-ef79-4a99-a546-d0bf1ffb3413.png)

**Reply by diazandr3s** [reply]
Sorry for the confusion ... I just wanted to provide the lines you should add in the Segmentation App.

Please add these lines in the segmentation App and let us know how that goes ...

![lines2add](https://user-images.githubusercontent.com/11991079/160005690-f6e90f92-e744-477b-8e62-821d880c5c1e.png)

**Reply by Suchi97** [reply]
Thanks @diazandr3s  for your reply. 
I tried but there is no difference from the above added predicted image.
![image](https://user-images.githubusercontent.com/25792114/160079124-3f2d4cc5-8ffb-4c6b-ab16-c040ced62f5b.png)

**Reply by diazandr3s** [reply]
You're using the BTCV dataset, right? 

Sorry, I now remembered we had similar issues with the image headers. This dataset is upside down. 

My suggestion is to use the itkreader to load the images in both infer and train:

> LoadImaged(keys="image", reader="ITKReader"),

I've also reoriented these images. You can download them from [here](https://emckclac-my.sharepoint.com/:f:/g/personal/k2039747_kcl_ac_uk/Eii_SC7CHh1Pm4hiSMVf8cUBBN1GgivMu8tRW32_0ejEVw?e=4nbQkU)

Please let me know how that goes.

**Reply by Suchi97** [reply]
Thanks @diazandr3s for your information.
Yes we are using BTCV dataset. 
We will try this and let you know the results

**Reply by Suchi97** [reply]
@diazandr3s , 

Using the dataset (BTCV_reoriented), I inferred the image img0002.nii.gz using MONAI Label Segmentation app and without MONAI Label

The below results are obtained without using MONAI Label

![image](https://user-images.githubusercontent.com/25792114/160157185-e3b7f035-916f-4f32-b46c-3300e0e7b260.png)

The below results are obtained when using MONAI Label Segmentation App
![image](https://user-images.githubusercontent.com/25792114/160157087-e67bf0c3-6e21-43ab-95bd-d6c60e32c302.png)

From the above image, we can some progress. But the labels are not mapped correctly. 

Can you please look into this issue.
Thanks

**Reply by diazandr3s** [reply]
Hi @Suchi97,

If you're using the **BTCV_reoriented** dataset, why the image is still upside down? I see the axial view isn't properly oriented.
I checked myself and got the following result:

![img002](https://user-images.githubusercontent.com/11991079/160212221-8ff1bdba-bc93-4410-b46b-426f5217d2d9.png)

Are you using an orientation transform? [These](https://github.com/Project-MONAI/MONAILabel/blob/main/sample-apps/radiology/lib/infers/segmentation.py#L55) are the ones I used. 

It seems there is an issue with a transformation. 

BTW, after checking in more detail, now I think you don't need to use the method I recommended you before:

![lines2add](https://user-images.githubusercontent.com/11991079/160212630-aba99676-482b-49fb-b68a-137ca6ca958e.png)

This is because there isn't transform to invert. Please remove those lines :)

I'd suggest you check why the images aren't properly oriented. Or maybe there is a transform that is not being applied in inference.

Keep us posted.

**Reply by Suchi97** [reply]
Thanks @diazandr3s  for your reply. It was really helpful.

We could able to solve the issue. When we checked with the image `img0031.nii.gz`, we got the following results

The image below shows the inference without using MONAI Label

![image](https://user-images.githubusercontent.com/25792114/161083753-44fd4e8c-0c3e-47d0-9efe-4786ed5692fc.png)

The image below shows the inference using the segmentation app in MONAI Label

![image](https://user-images.githubusercontent.com/25792114/161083837-6a6ae653-c5dd-4869-b27d-71382cc94bf8.png)

Now, we could able to use our own trained model in MONAI Label (multilabel). *[The model is not that good and the mean dice obtained for the above model is 0.89]*

Can you please help us to understand the reason behind the reorientation of BTCV images and how to reorient the images.
If we come across any other images like that, it will be helpful.

Thanks in advance.

**Reply by SachidanandAlle** [reply]
to answer your question on re-orientation.. you have to invert if you are doing it as part of pre-transform.

For example.. for deepedit, you have pretransform: `Orientationd(keys="image", axcodes="RAS")`
https://github.com/Project-MONAI/MONAILabel/blob/main/sample-apps/radiology/lib/infers/deepedit.py#L73

so you need to revert it back.. through invert/post transform

**Comment by SachidanandAlle** [comment]
I suggest to refer latest radiology app and there segmentation model..
https://github.com/Project-MONAI/MONAILabel/tree/main/sample-apps/radiology#segmentation

Similarly you can add your 3-label model into radiology app..
More details on how to add a new model:
https://github.com/Project-MONAI/MONAILabel/tree/main/sample-apps/radiology#how-to-add-new-model

few things are simplified new latest main branch (which shall be released in 0.4 version)

**Comment by diazandr3s** [comment]
Thanks for the update, @Suchi97. 
You could use the BTCV dataset as is to train DeepEdit or Segmentation App. 
The reason we've decided to reorient it first is because clinicians are used to see the images in this orientation:

![orientation](https://user-images.githubusercontent.com/11991079/161430764-dfa5b75b-c0af-4887-985b-e7b95dda3459.png)

If you don't mind having the images upside down, you don't need to reorient them first.

As @SachidanandAlle commented, all radiology apps were merged into [single App.](https://github.com/Project-MONAI/MONAILabel/tree/main/sample-apps/radiology)

Hope this helps.

**Reply by Suchi97** [reply]
Thanks @SachidanandAlle  and @diazandr3s for your reply. 

We could able to understood how to use our own trained model in MONAI Label.
We will also try with the new MONAI Label version also.

Thanks again :)

**Comment by DAC-ACD** [comment]
Hello! I have a similar questions.

Some information about my requests: The task is to make the automated segmentation, but I can't see my target organ in default labels (rectum and colon). So, I decided to train custom model for my needs, based on Your deepedit model. 

 I'd like to train deepedit model for several times with different volumes. As I understand, after every training session server downloading trained weights: 
"...\radiology\model\deepedit_dynunet.pt"
But also I see some more created weights:
"...\radiology\model\deepedit_dynunet\train_01\checkpoint_final.pt"
"...\radiology\model\deepedit_dynunet\train_01\model.pt"

So, the _first question_ is: which one is needed to continue training my new model?
The _second question_ is: if I continue training after going to the "next sample", I continue training of previous model or I creating a new one?
The _third question_ is: if the answer of the _second question_ "continue training of previous model", how can I start next session with previously trained model?

What I already understood: 
1) I need to enable pre-trained model:
`if strtobool(self.conf.get("use_pretrained_model", "true")):
            url = f"{self.conf.get('pretrained_path', self.PRE_TRAINED_PATH)}/deepedit_{network}_multilabel.pt"
            download_file(url, self.path[0])`  ->  "false" in deepedit.py
2) correct `self.labels` in deepedit.py

Also, that would be great, if You give me some more recommendations to simplify my work!

**Reply by diazandr3s** [reply]
> Hello! I have a similar questions.
> 
> Some information about my requests: The task is to make the automated segmentation, but I can't see my target organ in default labels (rectum and colon). So, I decided to train custom model for my needs, based on Your deepedit model.
> 
> I'd like to train deepedit model for several times with different volumes. As I understand, after every training session server downloading trained weights: "...\radiology\model\deepedit_dynunet.pt" But also I see some more created weights: "...\radiology\model\deepedit_dynunet\train_01\checkpoint_final.pt" "...\radiology\model\deepedit_dynunet\train_01\model.pt"
> 
> So, the _first question_ is: which one is needed to continue training my new model? The _second question_ is: if I continue training after going to the "next sample", I continue training of previous model or I creating a new one? The _third question_ is: if the answer of the _second question_ "continue training of previous model", how can I start next session with previously trained model?
> 
> What I already understood:
> 
>     1. I need to enable pre-trained model:
>        `if strtobool(self.conf.get("use_pretrained_model", "true")): url = f"{self.conf.get('pretrained_path', self.PRE_TRAINED_PATH)}/deepedit_{network}_multilabel.pt" download_file(url, self.path[0])`  ->  "false" in deepedit.py
> 
>     2. correct `self.labels` in deepedit.py
> 
> 
> Also, that would be great, if You give me some more recommendations to simplify my work!

Hi @DAC-ACD,


> So, the first question is: which one is needed to continue training my new model?

This one `"...\radiology\model\deepedit_dynunet.pt"`

> The second question is: if I continue training after going to the "next sample", I continue training of previous model or I creating a new one?

By default, it'll use the previous as starting point. It means it'll re-train the previous model. Unless you disable the flag pretrained in Slicer.

> The third question is: if the answer of the second question "continue training of previous model", how can I start next session with previously trained model?


If this file `"...\radiology\model\deepedit_dynunet.pt"`exists and the pretrained flag is enabled, you'll use that model.

This file gets updated automatically if during training there is a model performing better.
